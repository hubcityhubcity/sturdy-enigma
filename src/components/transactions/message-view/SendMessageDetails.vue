<template>
  <div>
    <div class="tx__content__caption">
      <p>
        {{ caption }}
        <b> {{ coin.amount | atoms | prettyLong }}</b>
        <span> {{ coin.denom | viewDenom }}</span>
      </p>
    </div>
    <div class="tx__content__information">
      <template v-if="toYourself"
        >To yourself!</template
      >
      <template v-else-if="sentFromSessionAddress">
        To&nbsp;
        <Bech32 :address="transaction.value.to_address" />
      </template>
      <template v-else-if="receivedToSessionAddress">
        From&nbsp;
        <Bech32 :address="transaction.value.from_address" />
      </template>
      <template v-else>
        From&nbsp;
        <Bech32 :address="transaction.value.from_address" />&nbsp;to&nbsp;
        <Bech32 :address="transaction.value.to_address" />
      </template>
      <span v-if="transaction.memo">&nbsp;- {{ transaction.memo }}</span>
    </div>
  </div>
</template>

<script>
import { atoms, viewDenom, prettyLong } from "scripts/num.js"
import Bech32 from "common/Bech32"
import { getCoin } from "scripts/transaction-utils"

export default {
  name: `send-message-details`,
  filters: {
    atoms,
    viewDenom,
    prettyLong
  },
  components: {
    Bech32
  },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    sessionAddress: {
      type: String,
      required: false,
      default: null
    }
  },
  computed: {
    coin() {
      return getCoin(this.transaction)
    },
    toYourself() {
      const value = this.transaction.value
      return (
        value.from_address === this.sessionAddress &&
        value.to_address === this.sessionAddress
      )
    },
    sentFromSessionAddress() {
      const value = this.transaction.value
      return (
        this.sessionAddress === value.from_address &&
        this.sessionAddress !== value.to_address
      )
    },
    receivedToSessionAddress() {
      const value = this.transaction.value
      return (
        this.sessionAddress === value.to_address &&
        this.sessionAddress !== value.from_address
      )
    },
    caption() {
      const value = this.transaction.value
      if (
        value.to_address === this.sessionAddress &&
        value.from_address !== this.sessionAddress
      ) {
        return "Received"
      } else {
        return "Sent"
      }
    }
  }
}
</script>
