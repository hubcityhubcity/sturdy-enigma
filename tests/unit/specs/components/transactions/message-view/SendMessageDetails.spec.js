import { shallowMount } from "@vue/test-utils"
import SendMessageDetails from "src/components/transactions/message-view/SendMessageDetails"

describe(`SendMessageDetails`, () => {
  let wrapper
  // Examples @ https://stargate.cosmos.network/txs?action=send
  const tx = {
    height: 2000000,
    value: {
      from_address: `cosmos1`,
      to_address: `cosmos1ahtlr29s38w23xxq7slcwmmz4c8x9efmr8qmee`,
      amount: [
        {
          amount: "10000000000",
          denom: "uatom"
        }
      ]
    },
    fee: {
      amount: "37",
      denom: "uatom"
    },
    hash: `ABCD1234`,
    group: `banking`,
    key: "keyhash",
    memo: "(Sent via Lunie)",
    timestamp: 123456789,
    liquidDate: null,
    type: `cosmos-sdk/MsgSend`
  }

  it(`renders a sent transaction message`, () => {
    wrapper = shallowMount(SendMessageDetails, {
      propsData: {
        transaction: tx,
        sessionAddress: "cosmos1"
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`renders a received transaction message`, () => {
    wrapper = shallowMount(SendMessageDetails, {
      propsData: {
        transaction: tx,
        sessionAddress: "cosmos1ahtlr29s38w23xxq7slcwmmz4c8x9efmr8qmee"
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
