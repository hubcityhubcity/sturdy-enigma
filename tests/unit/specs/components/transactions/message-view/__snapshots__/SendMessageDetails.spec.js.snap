// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessageDetails renders a received transaction message 1`] = `
<div
  class="tx__content"
>
  <transactionicon-stub
    transactiongroup="banking"
    transactiontype="Received"
  />
   
  <div
    class="tx__content__left"
  >
    <h3>
      Received
    </h3>
     
    <span>
      From 
    </span>
     
    <bech32-stub
      address="cosmos1"
    />
  </div>
   
  <div
    class="tx__content__right"
  >
    <p
      class="amount"
    >
      
      10,000 ATOM
    
    </p>
  </div>
</div>
`;

exports[`SendMessageDetails renders a sent transaction message 1`] = `
<div
  class="tx__content"
>
  <transactionicon-stub
    transactiongroup="banking"
    transactiontype="Sent"
  />
   
  <div
    class="tx__content__left"
  >
    <h3>
      Sent
    </h3>
     
    <span>
      To 
    </span>
     
    <bech32-stub
      address="cosmos1ahtlr29s38w23xxq7slcwmmz4c8x9efmr8qmee"
    />
  </div>
   
  <div
    class="tx__content__right"
  >
    <p
      class="amount"
    >
      
      10,000 ATOM
    
    </p>
  </div>
</div>
`;
